Add queue component

Add one queue component into the topology to test Queue RTEMS implementation within the OSAL.

Issue when linking the executable:

```sh
[100%] Linking CXX executable bin/leon3-rtems5/App
/home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libFw_CompQueued.a(QueuedComponentBase.cpp.obj): in function `_ZN2Fw19QueuedComponentBaseC2EPKc':
/home/musubi/codelab/fprime-rtems-poc/fprime/Fw/Comp/QueuedComponentBase.cpp:10: undefined reference to `_ZN2Os5QueueC1Ev'
/home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libFw_CompQueued.a(QueuedComponentBase.cpp.obj): in function `_ZN2Fw19QueuedComponentBaseD2Ev':
/home/musubi/codelab/fprime-rtems-poc/fprime/Fw/Comp/QueuedComponentBase.cpp:14: undefined reference to `_ZN2Os5QueueD1Ev'
/home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libFw_CompQueued.a(ActiveComponentBase.cpp.obj): in function `_ZN2Fw19ActiveComponentBaseC2EPKc':
/home/musubi/codelab/fprime-rtems-poc/fprime/Fw/Comp/ActiveComponentBase.cpp:33: undefined reference to `_ZN2Os4TaskC1Ev'
/home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libFw_CompQueued.a(ActiveComponentBase.cpp.obj): in function `_ZN2Fw19ActiveComponentBaseD2Ev':
/home/musubi/codelab/fprime-rtems-poc/fprime/Fw/Comp/ActiveComponentBase.cpp:37: undefined reference to `_ZN2Os4TaskD1Ev'
/home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libFw_CompQueued.a(ActiveComponentBase.cpp.obj): in function `_ZN2Fw19ActiveComponentBase5startEjjjj':
/home/musubi/codelab/fprime-rtems-poc/fprime/Fw/Comp/ActiveComponentBase.cpp:77: undefined reference to `_ZN2Os4Task5startERKN2Fw10StringBaseEPFvPvES5_jjjj'
/home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libFw_CompQueued.a(ActiveComponentBase.cpp.obj): in function `_ZN2Fw19ActiveComponentBase4joinEPPv':
/home/musubi/codelab/fprime-rtems-poc/fprime/Fw/Comp/ActiveComponentBase.cpp:91: undefined reference to `_ZN2Os4Task4joinEPPv'
/home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: /home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: DWARF error: could not find variable specification at offset 66d
/home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: DWARF error: could not find variable specification at offset 67a
lib/leon3-rtems5/libOs.a(QueueCommon.cpp.obj): in function `_ZN2Os5Queue4sendERKN2Fw19SerializeBufferBaseEiNS0_13QueueBlockingE':
/home/musubi/codelab/fprime-rtems-poc/fprime/Os/QueueCommon.cpp:18: undefined reference to `_ZN2Os5Queue4sendEPKhiiNS0_13QueueBlockingE'
/home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libOs.a(QueueCommon.cpp.obj): in function `_ZN2Os5Queue7receiveERN2Fw19SerializeBufferBaseERiNS0_13QueueBlockingE':
/home/musubi/codelab/fprime-rtems-poc/fprime/Os/QueueCommon.cpp:28: undefined reference to `_ZN2Os5Queue7receiveEPhiRiS2_NS0_13QueueBlockingE'
/home/musubi/rcc-1.3.0-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libOs.a(QueueCommon.cpp.obj): in function `_ZN2Os5Queue6createERKN2Fw10StringBaseEii':
/home/musubi/codelab/fprime-rtems-poc/fprime/Os/QueueCommon.cpp:44: undefined reference to `_ZN2Os5Queue14createInternalERKN2Fw10StringBaseEii'
collect2: error: ld returned 1 exit status
gmake[3]: *** [CMakeFiles/App.dir/build.make:112: bin/leon3-rtems5/App] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:1860: CMakeFiles/App.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:1867: CMakeFiles/App.dir/rule] Error 2
gmake: *** [Makefile:196: App] Error 2
[ERROR] CMake erred with return code 2
```

Awesome, thanks a lot for the feedback!

Arf, if I add `register_fprime_module(Os)` I cannot generate the cache anymore:

```sh
CMake Error at /home/musubi/codelab/fprime-rtems-poc/fprime/cmake/module.cmake:31 (add_library):
  add_library cannot create target "Os" because another target with the same
  name already exists.  The existing target is a static library created in
  source directory "/home/musubi/codelab/fprime-rtems-poc/fprime/Os".  See
  documentation for policy CMP0002 for more details.
Call Stack (most recent call first):
  /home/musubi/codelab/fprime-rtems-poc/fprime/cmake/module.cmake:80 (generate_base_module_properties)
  /home/musubi/codelab/fprime-rtems-poc/fprime/cmake/API.cmake:163 (generate_library)
  /home/musubi/codelab/fprime-rtems-poc/fprime-rtems/rtems/CMakeLists.txt:62 (register_fprime_module)


CMake Error at /home/musubi/codelab/fprime-rtems-poc/fprime/cmake/prescan.cmake:127 (message):
-- Including library /home/musubi/codelab/fprime-rtems-poc/fprime-rtems with manifests fprime-rtems.cmake
-- Performing CMake source prescan
-- Configuring incomplete, errors occurred!
  Failed to run prescan, cannot continue.
See also "/home/musubi/codelab/fprime-rtems-poc/App/build-fprime-automatic-leon3-rtems5/CMakeFiles/CMakeOutput.log".
Call Stack (most recent call first):
  /home/musubi/codelab/fprime-rtems-poc/fprime/cmake/FPrime.cmake:64 (perform_prescan)
  CMakeLists.txt:13 (include)
```

I don't see `register_fprime_module(Os)` in the other OSAL nodes (VxWorks, ThreadX, or FreeRTOS) - this seems odd to me.

I tried to add the "Dependency Melding" in CMake: https://github.com/ThibFrgsGmz/fprime-rtems-poc/blob/43226d51206e8c80f71bc9b230e0ed49233e139e/fprime-rtems/rtems/CMakeLists.txt#L43-L47

I have this error at the end:

```sh
-- Adding Deployment: App
-- Configuring done (9.6s)
CMake Error: The inter-target dependency graph contains the following strongly connected component (cycle):
  "Os" of type STATIC_LIBRARY
    depends on "fprime-rtems_rtems" (weak)
    depends on "fprime-rtems_rtems" (strong)
  "fprime-rtems_rtems" of type STATIC_LIBRARY
    depends on "Os" (weak)
    depends on "Os" (strong)
The component contains at least one cycle consisting of strong dependencies (created by add_dependencies) that cannot be broken.
CMake Generate step failed.  Build files cannot be regenerated correctly.
[ERROR] CMake erred with return code 1. Partial build cache remains. Run purge to clean-up.
```

If I put it in comment and just keep the `register_fprime_module(fprime-rtems_rtems)`, I have the error:

```sh
Completed generating HTML event tables...
[ 96%] Building CXX object App/Top/CMakeFiles/App_Top.dir/App.cpp.obj
[ 96%] Building CXX object App/Top/CMakeFiles/App_Top.dir/AppTopologyAc.cpp.obj
[ 96%] Building CXX object App/Top/CMakeFiles/App_Top.dir/FppConstantsAc.cpp.obj
[100%] Linking CXX static library ../../lib/leon3-rtems5/libApp_Top.a
[100%] Built target App_Top
[100%] Building CXX object CMakeFiles/App.dir/Main.cpp.obj
[100%] Linking CXX executable bin/leon3-rtems5/App
/home/life/rcc-1.3.1-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libFw_CompQueued.a(QueuedComponentBase.cpp.obj): in function `_ZN2Fw19QueuedComponentBaseC2EPKc':
/home/life/dev/fprime-rtems-poc/fprime/Fw/Comp/QueuedComponentBase.cpp:10: undefined reference to `_ZN2Os5QueueC1Ev'
/home/life/rcc-1.3.1-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libFw_CompQueued.a(QueuedComponentBase.cpp.obj): in function `_ZN2Fw19QueuedComponentBaseD2Ev':
/home/life/dev/fprime-rtems-poc/fprime/Fw/Comp/QueuedComponentBase.cpp:14: undefined reference to `_ZN2Os5QueueD1Ev'
/home/life/rcc-1.3.1-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: /home/life/rcc-1.3.1-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: DWARF error: could not find variable specification at offset 66d
/home/life/rcc-1.3.1-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: DWARF error: could not find variable specification at offset 67a
lib/leon3-rtems5/libOs.a(QueueCommon.cpp.obj): in function `_ZN2Os5Queue4sendERKN2Fw19SerializeBufferBaseEiNS0_13QueueBlockingE':
/home/life/dev/fprime-rtems-poc/fprime/Os/QueueCommon.cpp:18: undefined reference to `_ZN2Os5Queue4sendEPKhiiNS0_13QueueBlockingE'
/home/life/rcc-1.3.1-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libOs.a(QueueCommon.cpp.obj): in function `_ZN2Os5Queue7receiveERN2Fw19SerializeBufferBaseERiNS0_13QueueBlockingE':
/home/life/dev/fprime-rtems-poc/fprime/Os/QueueCommon.cpp:28: undefined reference to `_ZN2Os5Queue7receiveEPhiRiS2_NS0_13QueueBlockingE'
/home/life/rcc-1.3.1-gcc/bin/../lib/gcc/sparc-gaisler-rtems5/10.2.0/../../../../sparc-gaisler-rtems5/bin/ld: lib/leon3-rtems5/libOs.a(QueueCommon.cpp.obj): in function `_ZN2Os5Queue6createERKN2Fw10StringBaseEii':
/home/life/dev/fprime-rtems-poc/fprime/Os/QueueCommon.cpp:44: undefined reference to `_ZN2Os5Queue14createInternalERKN2Fw10StringBaseEii'
collect2: error: ld returned 1 exit status
gmake[3]: *** [CMakeFiles/App.dir/build.make:113: bin/leon3-rtems5/App] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:1861: CMakeFiles/App.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:1868: CMakeFiles/App.dir/rule] Error 2
gmake: *** [Makefile:196: App] Error 2
[ERROR] CMake erred with return code 2
```

--- a/.clangd
+++ b/.clangd
@@ -1,0 +1,2 @@
+CompileFlags:
+  CompilationDatabase: App/build-fprime-automatic-leon3-rtems5/       # Search build/ directory for compile_commands.json

--- a/App/App.cmake
+++ b/App/App.cmake
@@ -1,3 +1,4 @@
 add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/Ports/")
 add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/ComponentA/")
 add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/ComponentB/")
+add_fprime_subdirectory("${CMAKE_CURRENT_LIST_DIR}/ComponentC/")

--- a/App/ComponentA/ComponentA.cpp
+++ b/App/ComponentA/ComponentA.cpp
@@ -47,7 +47,7 @@ namespace App {
         F32 val
     )
   {
-    //printf("Hello world Component 1 | input= %f\n", val);
+    printf("Hello world Component A | input= %f\n", val);
     this->out_f32_out(0, val);
   }
 

--- a/App/ComponentB/ComponentB.cpp
+++ b/App/ComponentB/ComponentB.cpp
@@ -46,7 +46,8 @@ namespace App {
         F32 val
     )
   {
-    //printf("Hello world Component 2 | input= %f\n", val);
+    printf("Hello world Component B | input= %f\n", val);
+    this->out_f32_out(0, val);
   }
 
 } // end namespace App

--- a/App/ComponentC/CMakeLists.txt
+++ b/App/ComponentC/CMakeLists.txt
@@ -1,0 +1,7 @@
+# Register the standard build
+set(SOURCE_FILES
+  "${CMAKE_CURRENT_LIST_DIR}/ComponentC.cpp"
+  "${CMAKE_CURRENT_LIST_DIR}/ComponentC.fpp"
+)
+register_fprime_module()
+

--- a/App/ComponentC/ComponentC.cpp
+++ b/App/ComponentC/ComponentC.cpp
@@ -1,0 +1,53 @@
+// ======================================================================
+// \title  ComponentC.cpp
+// \author musubi
+// \brief  cpp file for ComponentC component implementation class
+// ======================================================================
+
+
+#include <App/ComponentC/ComponentC.hpp>
+#include <FpConfig.hpp>
+
+namespace App {
+
+  // ----------------------------------------------------------------------
+  // Construction, initialization, and destruction
+  // ----------------------------------------------------------------------
+
+  ComponentC ::
+    ComponentC(
+        const char *const compName
+    ) : ComponentCComponentBase(compName)
+  {
+
+  }
+
+  void ComponentC ::
+    init(
+        const NATIVE_INT_TYPE queueDepth,
+        const NATIVE_INT_TYPE instance
+    )
+  {
+    ComponentCComponentBase::init(queueDepth, instance);
+  }
+
+  ComponentC ::
+    ~ComponentC()
+  {
+
+  }
+
+  // ----------------------------------------------------------------------
+  // Handler implementations for user-defined typed input ports
+  // ----------------------------------------------------------------------
+
+  void ComponentC ::
+    in_f32_handler(
+        const NATIVE_INT_TYPE portNum,
+        F32 val
+    )
+  {
+    // TODO
+  }
+
+} // end namespace App

--- a/App/ComponentC/ComponentC.cpp
+++ b/App/ComponentC/ComponentC.cpp
@@ -4,7 +4,6 @@
 // \brief  cpp file for ComponentC component implementation class
 // ======================================================================
 
-
 #include <App/ComponentC/ComponentC.hpp>
 #include <FpConfig.hpp>
 
@@ -47,7 +46,7 @@ namespace App {
         F32 val
     )
   {
-    // TODO
+    printf("Async Hello world Component C | input= %f\n", val);
   }
 
   void ComponentC ::
@@ -56,7 +55,7 @@ namespace App {
         F32 val
     )
   {
-    // TODO
+    printf("Sync Hello world Component C | input= %f\n", val);
   }
 
 } // end namespace App

--- a/App/ComponentC/ComponentC.cpp
+++ b/App/ComponentC/ComponentC.cpp
@@ -50,4 +50,13 @@ namespace App {
     // TODO
   }
 
+  void ComponentC ::
+    in_f32_2_handler(
+        const NATIVE_INT_TYPE portNum,
+        F32 val
+    )
+  {
+    // TODO
+  }
+
 } // end namespace App

--- a/App/ComponentC/ComponentC.fpp
+++ b/App/ComponentC/ComponentC.fpp
@@ -1,6 +1,7 @@
 module App {
-  active component ComponentC {
+  queued component ComponentC {
     async input port in_f32: prtF32
+    sync input port in_f32_2: prtF32
     output port out_f32: prtF32
   }
 }

--- a/App/ComponentC/ComponentC.fpp
+++ b/App/ComponentC/ComponentC.fpp
@@ -1,0 +1,6 @@
+module App {
+  active component ComponentC {
+    async input port in_f32: prtF32
+    output port out_f32: prtF32
+  }
+}

--- a/App/ComponentC/ComponentC.hpp
+++ b/App/ComponentC/ComponentC.hpp
@@ -51,6 +51,13 @@ namespace App {
           F32 val 
       );
 
+      //! Handler implementation for in_f32_2
+      //!
+      void in_f32_2_handler(
+          const NATIVE_INT_TYPE portNum, /*!< The port number*/
+          F32 val 
+      );
+
 
     };
 

--- a/App/ComponentC/ComponentC.hpp
+++ b/App/ComponentC/ComponentC.hpp
@@ -1,0 +1,59 @@
+// ======================================================================
+// \title  ComponentC.hpp
+// \author musubi
+// \brief  hpp file for ComponentC component implementation class
+// ======================================================================
+
+#ifndef ComponentC_HPP
+#define ComponentC_HPP
+
+#include "App/ComponentC/ComponentCComponentAc.hpp"
+
+namespace App {
+
+  class ComponentC :
+    public ComponentCComponentBase
+  {
+
+    public:
+
+      // ----------------------------------------------------------------------
+      // Construction, initialization, and destruction
+      // ----------------------------------------------------------------------
+
+      //! Construct object ComponentC
+      //!
+      ComponentC(
+          const char *const compName /*!< The component name*/
+      );
+
+      //! Initialize object ComponentC
+      //!
+      void init(
+          const NATIVE_INT_TYPE queueDepth, /*!< The queue depth*/
+          const NATIVE_INT_TYPE instance = 0 /*!< The instance number*/
+      );
+
+      //! Destroy object ComponentC
+      //!
+      ~ComponentC();
+
+    PRIVATE:
+
+      // ----------------------------------------------------------------------
+      // Handler implementations for user-defined typed input ports
+      // ----------------------------------------------------------------------
+
+      //! Handler implementation for in_f32
+      //!
+      void in_f32_handler(
+          const NATIVE_INT_TYPE portNum, /*!< The port number*/
+          F32 val 
+      );
+
+
+    };
+
+} // end namespace App
+
+#endif

--- a/App/Top/CMakeLists.txt
+++ b/App/Top/CMakeLists.txt
@@ -3,4 +3,11 @@ set(SOURCE_FILES
   "${CMAKE_CURRENT_LIST_DIR}/topology.fpp"
   "${CMAKE_CURRENT_LIST_DIR}/App.cpp"
 )
+
+# Note: supply non-explicit dependencies here
+set(MOD_DEPS
+  fprime-rtems/rtems
+  Os
+)
+
 register_fprime_module()

--- a/App/Top/CMakeLists.txt
+++ b/App/Top/CMakeLists.txt
@@ -4,10 +4,4 @@ set(SOURCE_FILES
   "${CMAKE_CURRENT_LIST_DIR}/App.cpp"
 )
 
-# Note: supply non-explicit dependencies here
-set(MOD_DEPS
-  fprime-rtems/rtems
-  Os
-)
-
 register_fprime_module()

--- a/App/Top/instances.fpp
+++ b/App/Top/instances.fpp
@@ -1,4 +1,15 @@
 module App {
+
+  module Default {
+    constant queueSize = 10
+    constant stackSize = 64 * 1024
+  }
+
+  instance componentC: App.ComponentC base id 0x0100 \
+    queue size Default.queueSize \
+    stack size Default.stackSize \
+    priority 100
+
   instance componentA: App.ComponentA base id 0x4A00
   instance componentB: App.ComponentB base id 0x4A00
 

--- a/App/Top/instances.fpp
+++ b/App/Top/instances.fpp
@@ -7,8 +7,8 @@ module App {
 
   instance componentC: App.ComponentC base id 0x0100 \
     queue size Default.queueSize \
-    stack size Default.stackSize \
-    priority 100
+    # stack size Default.stackSize \
+    # priority 100
 
   instance componentA: App.ComponentA base id 0x4A00
   instance componentB: App.ComponentB base id 0x4A00

--- a/App/Top/topology.fpp
+++ b/App/Top/topology.fpp
@@ -7,6 +7,8 @@ module App {
 
     connections ComponentBenchmark {
       componentA.out_f32 -> componentB.in_f32
+      #componentB.out_f32 -> componentC.in_f32
+      componentB.out_f32 -> componentC.in_f32_2
     }
 
   }

--- a/App/Top/topology.fpp
+++ b/App/Top/topology.fpp
@@ -3,6 +3,7 @@ module App {
   topology App {
     instance componentA
     instance componentB
+    instance componentC
 
     connections ComponentBenchmark {
       componentA.out_f32 -> componentB.in_f32

--- a/README.md
+++ b/README.md
@@ -14,6 +14,8 @@ F´ PoC on RTEMS LEON 3 w/ Renode
 
 # Standalone test
 
+## Renode
+
 - Install latest Renode: `wget https://github.com/renode/renode/releases/download/v1.13.3/renode-1.13.3.linux-portable.tar.gz`
 
 - Extract the archve `tar -xvzf renode-1.13.3.linux-portable.tar.gz`
@@ -24,6 +26,22 @@ RENODE_PATH='/home/user/renode_1.13.3_portable'
 export PATH="$RENODE_PATH:$PATH"
 ```
 
+## Gailer toolsuite
+### bcc2
+
+```
+curl -LO 'https://www.gaisler.com/anonftp/bcc2/bin/bcc-2.2.0-gcc-linux64.tar.xz'
+tar xf bcc-2.2.0-gcc-linux64.tar.xz
+```
+
+### grlib
+
+```
+curl -LO 'https://www.gaisler.com/products/grlib/grlib-gpl-2021.2-b4267.tar.gz'
+tar xf grlib-gpl-2021.2-b4267.tar.gz
+```
+
+# Usage
 - Launch renode and the Renode script `.resc`:
 
 ```sh

--- a/fprime-rtems/cmake/toolchain/leon3-rtems5.cmake
+++ b/fprime-rtems/cmake/toolchain/leon3-rtems5.cmake
@@ -13,7 +13,7 @@ set(CMAKE_TRY_COMPILE_TARGET_TYPE
     CACHE STRING "Try Static Lib Type" FORCE)
 
 # Adapt the path according to the developer setup, it seems that only absolute paths work
-set(TOOLCHAIN_PATH "/home/musubi/rcc-1.3.0-gcc" CACHE PATH "Toolchain path")
+set(TOOLCHAIN_PATH "/home/life/rcc-1.3.1-gcc" CACHE PATH "Toolchain path")
 
 # Check toolchain directory exists
 IF(NOT EXISTS "${TOOLCHAIN_PATH}")

--- a/fprime-rtems/rtems/CMakeLists.txt
+++ b/fprime-rtems/rtems/CMakeLists.txt
@@ -19,6 +19,5 @@ list(APPEND SOURCE_FILES
     "${FPRIME_FRAMEWORK_PATH}/Os/FileCommon.cpp"
 )
 
-register_fprime_module(fprime-rtems_rtems)
-
-target_link_libraries(Os PRIVATE fprime-rtems_rtems)
+register_fprime_module(fprime_rtems)
+target_link_libraries(Os PRIVATE fprime_rtems)

--- a/fprime-rtems/rtems/CMakeLists.txt
+++ b/fprime-rtems/rtems/CMakeLists.txt
@@ -1,25 +1,3 @@
-
-# set(MOD_DEPS
-#         "${CMAKE_THREAD_LIBS_INIT}"
-#         Fw/Cfg
-#         Fw/Types
-#         Fw/Logger
-#         Utils/Hash
-#         )
-
-# set(SOURCE_FILES
-#     "Event.cpp"
-#     "InterruptLock.cpp"
-#     "Log.cpp"
-#     "Mutex.cpp"
-#     "Queue.cpp"
-#     "Task.cpp"
-#     "TaskId.cpp"
-#     "TaskLock.cpp"
-#     "WatchdogTimerCommon.cpp"
-#     "${FPRIME_FRAMEWORK_PATH}/Os/Posix/IntervalTimer.cpp"
-#     )
-
 set(SOURCE_FILES
     "Mutex.cpp"
     "Queue.cpp"
@@ -41,4 +19,8 @@ elseif(CMAKE_SYSTEM_VERSION MATCHES 5.[0-9])
     message(SYSTEM VERSION ==== > 5.x)
 endif()
 
-register_fprime_module()
+set(MOD_DEPS
+    Os
+)
+
+register_fprime_module(fprime-rtems_rtems)

--- a/fprime-rtems/rtems/CMakeLists.txt
+++ b/fprime-rtems/rtems/CMakeLists.txt
@@ -19,6 +19,21 @@ elseif(CMAKE_SYSTEM_VERSION MATCHES 5.[0-9])
     message(SYSTEM VERSION ==== > 5.x)
 endif()
 
+# Basic source files used in every OSAL layer.  Contains common code and helpers.
+set(SOURCE_FILES
+    "${FPRIME_FRAMEWORK_PATH}/Os/IntervalTimerCommon.cpp"
+    "${FPRIME_FRAMEWORK_PATH}/Os/TaskString.cpp"
+    "${FPRIME_FRAMEWORK_PATH}/Os/TaskCommon.cpp"
+    "${FPRIME_FRAMEWORK_PATH}/Os/QueueCommon.cpp"
+    "${FPRIME_FRAMEWORK_PATH}/Os/QueueString.cpp"
+    "${FPRIME_FRAMEWORK_PATH}/Os/IPCQueueCommon.cpp"
+    "${FPRIME_FRAMEWORK_PATH}/Os/SimpleQueueRegistry.cpp"
+    "${FPRIME_FRAMEWORK_PATH}/Os/MemCommon.cpp"
+    "${FPRIME_FRAMEWORK_PATH}/Os/ValidateFileCommon.cpp"
+    "${FPRIME_FRAMEWORK_PATH}/Os/ValidatedFile.cpp"
+    "${FPRIME_FRAMEWORK_PATH}/Os/FileCommon.cpp"
+)
+
 set(MOD_DEPS
     Os
 )

--- a/fprime-rtems/rtems/CMakeLists.txt
+++ b/fprime-rtems/rtems/CMakeLists.txt
@@ -5,22 +5,7 @@ set(SOURCE_FILES
     "TaskId.cpp"
     )
 
-if(CMAKE_SYSTEM_VERSION MATCHES 4.[789])
-#   list(APPEND SOURCE_FILES
-#         "4.x/Task.cpp"
-#         "4.x/WatchdogTimer.cpp"
-#       )
-    message(SYSTEM VERSION ==== > 4.x)
-elseif(CMAKE_SYSTEM_VERSION MATCHES 5.[0-9])
-#   list(APPEND SOURCE_FILES
-#         "5.x/Task.cpp"
-#         "4.x/WatchdogTimer.cpp"
-#       )
-    message(SYSTEM VERSION ==== > 5.x)
-endif()
-
-# Basic source files used in every OSAL layer.  Contains common code and helpers.
-set(SOURCE_FILES
+list(APPEND SOURCE_FILES
     "${FPRIME_FRAMEWORK_PATH}/Os/IntervalTimerCommon.cpp"
     "${FPRIME_FRAMEWORK_PATH}/Os/TaskString.cpp"
     "${FPRIME_FRAMEWORK_PATH}/Os/TaskCommon.cpp"
@@ -34,14 +19,6 @@ set(SOURCE_FILES
     "${FPRIME_FRAMEWORK_PATH}/Os/FileCommon.cpp"
 )
 
-set(MOD_DEPS
-    Os
-)
-
 register_fprime_module(fprime-rtems_rtems)
 
-if (${CMAKE_SYSTEM_NAME} STREQUAL "RTEMS")
-    add_dependencies(Os fprime-rtems_rtems)
-    target_link_libraries(Os PRIVATE fprime-rtems_rtems)
-    target_link_libraries(fprime-rtems_rtems PRIVATE Os)
-endif()
+target_link_libraries(Os PRIVATE fprime-rtems_rtems)

--- a/fprime-rtems/rtems/CMakeLists.txt
+++ b/fprime-rtems/rtems/CMakeLists.txt
@@ -39,3 +39,9 @@ set(MOD_DEPS
 )
 
 register_fprime_module(fprime-rtems_rtems)
+
+if (${CMAKE_SYSTEM_NAME} STREQUAL "RTEMS")
+    add_dependencies(Os fprime-rtems_rtems)
+    target_link_libraries(Os PRIVATE fprime-rtems_rtems)
+    target_link_libraries(fprime-rtems_rtems PRIVATE Os)
+endif()

--- a/fprime-rtems/rtems/Mutex.cpp
+++ b/fprime-rtems/rtems/Mutex.cpp
@@ -14,22 +14,25 @@ namespace Os {
         rtems_status_code                status;
         rtems_name                       r_name;
         status = rtems_semaphore_create(r_name, 1, OSAL_MUTEX_ATTRIBS, 0, &this->m_handle);
-        FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+        // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
     }
     
     Mutex::~Mutex(void) {
+        rtems_status_code                status;
         status = rtems_semaphore_delete(this->m_handle);
-        FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+        // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
     }
     
     void Mutex::lock(void) {
+        rtems_status_code                status;
         status = rtems_semaphore_obtain(this->m_handle, RTEMS_WAIT, RTEMS_NO_TIMEOUT);
-        FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+        // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
     }
 
     void Mutex::unLock(void) {
+        rtems_status_code                status;
         status = rtems_semaphore_release(this->m_handle);
-        FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+        // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
     }
 }
 

--- a/fprime-rtems/rtems/Queue.cpp
+++ b/fprime-rtems/rtems/Queue.cpp
@@ -26,7 +26,7 @@ namespace Os {
 
         if (status != RTEMS_SUCCESSFUL)
         {
-            FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+            // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
             return QUEUE_UNINITIALIZED;
         }
 
@@ -37,7 +37,7 @@ namespace Os {
     Queue::~Queue() {
         rtems_status_code                status;
         status = rtems_message_queue_delete(this->m_handle);
-        FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+        // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
         // (void)msgQDelete((MSG_Q_ID) this->m_handle);
     }
 
@@ -54,9 +54,9 @@ namespace Os {
         rtems_status_code                status;
         rtems_id                         rtems_queue_id;
         rtems_queue_id = this->m_handle;
-        status = rtems_message_queue_send(rtems_queue_id, data, size);
+        status = rtems_message_queue_send(rtems_queue_id, buffer, size);
 
-        FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+        // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
 
         if (status != RTEMS_SUCCESSFUL) {
             switch (status) {
@@ -77,6 +77,7 @@ namespace Os {
 
         rtems_status_code status;
         rtems_interval ticks;
+        rtems_option      option_set;
         rtems_id rtems_queue_id = this->m_handle;
 
         if (block == QUEUE_NONBLOCKING) {
@@ -88,9 +89,9 @@ namespace Os {
             ticks      = RTEMS_NO_TIMEOUT;
         }
 
-        status = rtems_message_queue_receive(rtems_queue_id, buffer, &capacity, option_set, ticks);
+        status = rtems_message_queue_receive(rtems_queue_id, buffer, (size_t*) &capacity, option_set, ticks);
 
-        FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+        // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
 
         if (status != RTEMS_SUCCESSFUL) {
             capacity = 0;
@@ -108,10 +109,10 @@ namespace Os {
 
     NATIVE_INT_TYPE Queue::getNumMsgs(void) const {
 
-        uint32_t = number_pending_msg;
+        uint32_t number_pending_msg;
         rtems_status_code status = rtems_message_queue_get_number_pending(this->m_handle, &number_pending_msg);
 
-        FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+        // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
 
         if (status != RTEMS_SUCCESSFUL) {
             return -1;

--- a/fprime-rtems/rtems/Task.cpp
+++ b/fprime-rtems/rtems/Task.cpp
@@ -36,7 +36,7 @@ namespace Os {
 
         status = rtems_task_create(r_name, priority, stackSize, r_mode, r_attributes, &this->m_handle);
 
-        FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+        // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
 
         if (status != RTEMS_SUCCESSFUL)
         {
@@ -45,7 +45,7 @@ namespace Os {
         
         status = rtems_task_start(this->m_handle, (rtems_task_entry)routine, (rtems_task_argument)r_name);
 
-        FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
+        // FW_ASSERT(status == RTEMS_SUCCESSFUL, rtems_status_text(status));
 
         if (status != RTEMS_SUCCESSFUL)
         {
@@ -89,7 +89,7 @@ namespace Os {
     Task::TaskStatus Task::delay(NATIVE_UINT_TYPE milliseconds)
     {
         int   tick_count;
-        int32 return_code = TASK_OK;
+        int32_t return_code = TASK_OK;
         Task::TaskStatus status = TASK_DELAY_ERROR;
 
         // return_code = OS_Milli2Ticks(milli_second, &tick_count);
@@ -106,7 +106,7 @@ namespace Os {
         }
 
         rtems_task_wake_after((rtems_interval)tick_count);
-        return return_code;
+        return (Task::TaskStatus)return_code;
     }
 
     void Task::suspend(bool onPurpose) {

--- a/fprime-rtems/rtems/TaskId.cpp
+++ b/fprime-rtems/rtems/TaskId.cpp
@@ -6,7 +6,7 @@ namespace Os
 {
     TaskId::TaskId(void)
     {
-        id = taskIdSelf();
+        id = rtems_task_self();
     }
     TaskId::~TaskId(void)
     {

--- a/fprime-rtems/rtems/TaskId.cpp
+++ b/fprime-rtems/rtems/TaskId.cpp
@@ -1,5 +1,6 @@
 // File TaskId.cpp
 #include <rtems.h>
+#include "fprime-leon3/Types/PlatformTypes.hpp"
 #include <Os/TaskId.hpp>
 
 namespace Os

--- a/swil/build-rtems.sh
+++ b/swil/build-rtems.sh
@@ -2,7 +2,7 @@
 set -euo pipefail
 
 curl -LO https://www.gaisler.com/anonftp/rcc/bin/linux/sparc-rtems-5-gcc-10.2.0-1.3.1-linux.txz
-tar -Jxf sparc-rtems-5-gcc-10.2.0-1.3.0-linux.txz
+tar -Jxf sparc-rtems-5-gcc-10.2.0-1.3.1-linux.txz
 pushd rcc-1.3.0-gcc/src/samples/
 # Add rtems-shell application to LEON3 target
 # as by default it is included only for LEON2

--- a/swil/build-rtems.sh
+++ b/swil/build-rtems.sh
@@ -1,7 +1,7 @@
 #!/bin/bash
 set -euo pipefail
 
-curl -LO https://www.gaisler.com/anonftp/rcc/bin/linux/sparc-rtems-5-gcc-10.2.0-1.3.0-linux.txz
+curl -LO https://www.gaisler.com/anonftp/rcc/bin/linux/sparc-rtems-5-gcc-10.2.0-1.3.1-linux.txz
 tar -Jxf sparc-rtems-5-gcc-10.2.0-1.3.0-linux.txz
 pushd rcc-1.3.0-gcc/src/samples/
 # Add rtems-shell application to LEON3 target

--- a/swil/leon3_rtems.resc
+++ b/swil/leon3_rtems.resc
@@ -1,5 +1,8 @@
 :name: Leon3 rtems demo
 :description: This script runs the rtems shell demo on Leon3
+: Run
+:renode -P 1234 --disable-xwt --console
+:include @swil/leon3_rtems.resc
 
 using sysbus
 $name?="Leon3"
@@ -23,3 +26,5 @@ macro reset
 """
 
 runMacro $reset
+
+:start

--- a/swil/leon3_rtems.resc
+++ b/swil/leon3_rtems.resc
@@ -12,8 +12,11 @@ machine LoadPlatformDescription @platforms/boards/leon3.repl
 
 path add $ORIGIN
 
-$prom?="/home/musubi/codelab/fprime-rtems-poc/swil/grlib-gpl-2021.2-b4267/software/leon3/prom.bin"
-$bin?="/home/musubi/codelab/fprime-rtems-poc/App/build-artifacts/leon3-rtems5/bin/App"
+:$prom?="/home/musubi/codelab/fprime-rtems-poc/swil/grlib-gpl-2021.2-b4267/software/leon3/prom.bin"
+:$bin?="/home/musubi/codelab/fprime-rtems-poc/App/build-artifacts/leon3-rtems5/bin/App"
+
+$prom="/home/life/grlib-gpl-2021.2-b4267/software/leon3/prom.bin"
+$bin="/home/life/dev/fprime-rtems-poc/App/build-artifacts/leon3-rtems5/bin/App"
 
 showAnalyzer sysbus.uart
 

--- a/swil/leon3_rtems_fprime.robot
+++ b/swil/leon3_rtems_fprime.robot
@@ -13,8 +13,8 @@ ${PROMPT}                     SHLL [/] #
 *** Keywords ***
 Run Fprime
     
-    Execute Command           $prom="${CURDIR}/grlib-gpl-2021.2-b4267/software/leon3/prom.bin"
-    Execute Command           $bin="/home/musubi/codelab/fprime-rtems-poc/App/build-artifacts/leon3-rtems5/bin/App"
+    Execute Command           $prom="/home/life/grlib-gpl-2021.2-b4267/software/leon3/prom.bin"
+    Execute Command           $bin="/home/life/dev/fprime-rtems-poc/App/build-artifacts/leon3-rtems5/bin/App"
     
     Execute Script            ${SCRIPT}
     Set Default Uart Timeout  2
@@ -28,5 +28,4 @@ Should Boot RTEMS with FPrime
 
     Start Emulation
 
-    Wait For Prompt On Uart   Hello Worldd
 

--- a/swil/test.sh
+++ b/swil/test.sh
@@ -10,13 +10,13 @@ TESTS_TO_RUN='leon3_rtems_fprime.robot'
 # python3 -m venv swil-venv
 # . ./swil-venv/bin/activate
 
-# # Fix 'error: invalid command 'bdist_wheel''
+# Fix 'error: invalid command 'bdist_wheel''
 # pip3 install wheel
 
-# # Install renode-run script
+# Install renode-run script
 # pip install -q git+https://github.com/antmicro/renode-run
 
-# # Download renode
+# Download renode
 # renode-run -a $RENODE_RUN_DIR download -d $RENODE_VERSION
 
 # Run test suite

